mesop/dataclass_utils/dataclass_utils.py

```diff
@@ -221,6 +238,18 @@ def decode_mesop_json_state_hook(dct):
     if _PANDAS_OBJECT_KEY in dct:
       return pd.read_json(StringIO(dct[_PANDAS_OBJECT_KEY]), orient="table")
 
+  if _PYDANTIC_OBJECT_KEY in dct:
+    cache_key = (
+      dct[_PYDANTIC_OBJECT_KEY]["module"],
+      dct[_PYDANTIC_OBJECT_KEY]["qualname"],
+    )
+    if cache_key not in pydantic_model_cache:
+      raise MesopException(
+        f"Tried to deserialize Pydantic model, but it's not in the cache: {cache_key}"
+      )
+    model_class = pydantic_model_cache[cache_key]
+    return model_class.model_validate_json(dct[_PYDANTIC_OBJECT_KEY]["json"])
```

Comment:
Does model_validate_json validate the model if it has validators on it -- or is it just the JSON? Could there be a case where the model has invalid data based on validator logic ?

Reply:
Also curious about the deserialization since it uses pydantic. Could it be assumed that it could handle some serialization errors that we currently run into (such as tuples, unions, etc).

Reply:
model_validate_json should validate with the validators - see: https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_validate_json.

Yeah - see the test cases I added, this is much more robust than what we currently support. I'm going to try using this in my personal project and if it turns out to be solid, I think we can add a recommendation in our [docs](https://google.github.io/mesop/guides/state-management/) to use Pydantic.